### PR TITLE
NMS-12092 add test to verify REST endpoint for search

### PR DESCRIPTION
**Pull request by patrick-schweizer**
### All Contributors

* [ X] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [ X] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>
  If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [X] Have you made a comment in that issue which points back to this PR?
* [X] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [ ] If this is a new or updated feature, is there documentation for the new behavior?
* [X] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

Once the reviewer(s) accept the PR and the branch passes continuous integration in Bamboo, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-12092
* Bamboo (Continuous Integration): https://bamboo.opennms.org/

### CODE REVIEW

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.restassured.RestAssured;
+import io.restassured.http.ContentType;
+import io.restassured.response.Response;
+
+public class GraphSearchRestServiceIT  extends OpenNMSSeleniumTestCase {
+
+	private final static Logger LOG = LoggerFactory.getLogger(GraphSearchRestServiceIT.class);
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: Formatting issue.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    @Before
+    public void setUp() {
+        RestAssured.baseURI = getBaseUrl();
+        RestAssured.port = getServerHttpPort();
+        RestAssured.basePath = "/opennms/api/v2/graphs";
+        RestAssured.authentication = preemptive().basic(BASIC_AUTH_USERNAME, BASIC_AUTH_PASSWORD);
+
+        // Install features
+        karafShell = new KarafShell(getServiceAddress(NewTestEnvironment.ContainerAlias.OPENNMS, 8101));
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: Not sure if this will work, as there is no container.

Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: it seems to work, it is the same code as in https://github.com/OpenNMS/opennms/blob/jira/NMS-12092-add-rest-test/smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java

Comment (mvrueden) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: Looking at both classes, I say we merge them, as they are duplicating code.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java (hunk 1 above)
Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: what is the issue?

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java (hunk 2 above)
Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: ok, done.

**Comment by patrick-schweizer** [comment]
@mvrueden I merged the two test classes.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java (hunk 1 above)
Comment (mvrueden) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: Screenshot 2019-06-19 at 18 03 06

The spacing is off.

Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphSearchRestServiceIT.java]: fixed.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java [hunk 3]
```diff
@@ -128,7 +132,65 @@ public void verifyGetContainer() {
...
+    	createGraphMLAndWaitUntilDone();
+        given().params("s", "unknown")
+               .accept(ContentType.JSON)
+               .get("/search/suggestions/{namespace}/", "acme:regions")
+               .then().statusCode(204);
+        
+        Response response = given().params("s", "North Region")
+               .accept(ContentType.JSON)
+               .get("/search/suggestions/{namespace}/", "acme:regions");
+        LOG.info("Results from search:" + response.asString());
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: You can use `RestAssured` for this, there is a `.log()`-statement: https://github.com/rest-assured/rest-assured/wiki/usage#request-logging

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java [hunk 4]
```diff
@@ -128,7 +132,65 @@ public void verifyGetContainer() {
...
+        
+        response
+               .then().statusCode(200)
+               .contentType(ContentType.JSON)
+               .content("[0].context", Matchers.is("GenericVertex"))
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: Let's also verify that the returned size of elements is correct.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java [hunk 5]
```diff
@@ -128,7 +132,65 @@ public void verifyGetContainer() {
...
+               .get("/search/results/{namespace}/", "acme:regions")
+               .then().statusCode(204);
+
+        Response response = given()
+               .params("providerId", "LabelSearchProvider")
+               .params("criteria", "North Region")
+               .params("context", "GenericVertex")
+               .accept(ContentType.JSON)
+               .get("/search/results/{namespace}/", "acme:regions");
+        LOG.info("Results from search:" + response.asString());
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: See above.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java [hunk 6]
```diff
@@ -128,7 +132,65 @@ public void verifyGetContainer() {
...
+        
+        response.then().statusCode(200)
+               .contentType(ContentType.JSON)
+               .content("[0].properties.namespace", Matchers.is("acme:regions"))
```
Comment (mvrueden) [inline smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: See above.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java (hunk 3 above)
Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: :+1:  done.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java (hunk 4 above)
Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: done.

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java (hunk 5 above)
Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: done

File: smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java (hunk 6 above)
Comment (patrick-schweizer) [reply smoke-test/src/test/java/org/opennms/smoketest/graph/GraphRestServiceIT.java]: done

**Comment by mvrueden** [comment]
Looks like circleci is now failing. I also merged latest `develop` to `features/graph-service`, so you should rebase in order to get this merged.

Let me know if you need any help.

**Pull request merged**